raft: introduce non-protobuf structs

andrewbaptist opened the pull request:
<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Commit:
raft: introduce non-protobuf structs

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131511)
<!-- Reviewable:end -->

Commit:
raft: automatic reformat

```

names=$(grep '^type' raft/rafttype/types.go  | awk '{print $2}')

for x in $(grep -l -r raftpb raft kv raftpb); do
	echo $x
        sed -i '' 's/pb "github.com\/cockroachdb\/cockroach\/pkg\/raft\/raftpb"/rt "github\.com\/cockroachdb\/cockroach\/pkg\/raft\/rafttype"/' $x
        sed -i '' 's/"github.com\/cockroachdb\/cockroach\/pkg\/raft\/raftpb"/"github\.com\/cockroachdb\/cockroach\/pkg\/raft\/rafttype"/' $x

        for name in $names Conf Msg; do
                sed -i '' "s/raftpb.$name/rafttype.$name/g" $x
                sed -i '' "s/pb.$name/rt.$name/g" $x
        done
done
```

Commit:
raft: Fixes as part of the raft type split

pav-kv commented on pkg/raft/rafttype/util.go:
```diff
@@ -0,0 +1,211 @@
+}
+
+func convertEntries(entries []raftpb.Entry) []Entry {
+	result := make([]Entry, len(entries))
+	for i, entry := range entries {
+		result[i] = Entry{
+			Term:  entry.Term,
+			Index: entry.Index,
+			Type:  EntryType(entry.Type),
+			Data:  entry.Data,
+		}
+	}
```
This causes an extra allocation on every message containing entries (`MsgProp`, `MsgApp`, etc)? Also, the `convertSnapshot`.

We would probably want to make these messages iterable or zero-copy somehow, so that they could be stepped into raft without an extra allocation.

pav-kv reviewed (commented):
Looks interesting. I don't know if we would want `rafttype.Message` to be a "union" of all the message types in the end, but seems like some starting point.

It would be nice to explore having a strong type for each message instead of a "union". Then, instead of `Step(Message)`, do something like: `StepMsgApp(MsgApp)`, `StepMsgAppResp(MsgAppResp)` - these would give the caller some error or introspection specific to this step rather than a generic error. This introspection would be useful to the upper layer, e.g. RACv2 would benefit knowing which range of entries got appended (an append can be partial).

There are some performance concerns here: conversions need to copy all the fields, and allocate in some cases (e.g. when there are `Entries` in the message). The copies aren't too bad, but the allocations can probably be worked around in some way.

The scope of this change can probably be reduced for the moment. What are we trying to achieve? If we are trying to annotate `pb.Entry` with in-memory information (trace context, etc), then we could probably do this with a smaller API change, or outside `raft`.
